src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -0,0 +1,658 @@
+                    }
+                }
+
+                // Returing a value other than size fails the callback and forces
+                // request completion with an error
+                return size - 1;
+            }
+
+            private unsafe static size_t CurlReceiveBodyCallback(
+                byte* buffer, size_t size, size_t nitems, IntPtr context)
+            {
+                size *= nitems;
+
+                EasyRequest easy;
+                if (TryGetEasyRequestFromContext(context, out easy))
+                {
+                    try
+                    {
+                        if (!(easy.Task.IsCanceled || easy.Task.IsFaulted))
+                        {
+                            // Complete the task if it hasn't already been.  This will make the
+                            // stream available to consumers.  A previous write callback
+                            // may have already completed the task to publish the response.
+                            easy.EnsureResponseMessagePublished();
+
+                            // Wait for a reader
+                            // TODO: The below call blocks till all the data has been read since
+                            //       response body is not suppored to be buffered in memory.
+                            //       Figure out some way to work around this
+                            if (size != 0)
+                            {
+                                easy.ResponseMessage.ContentStream.WaitAndSignalReaders(buffer, (long)size);
+                            }
+
+                            return size;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        easy.FailRequest(ex); // cleanup will be handled by main processing loop
+                    }
+                }
+
+                // Returing a value other than size fails the callback and forces
+                // request completion with an error.
+                return (size > 0) ? size - 1 : 1;
+            }
+
+            private static size_t CurlSendCallback(IntPtr buffer, size_t size, size_t nitems, IntPtr context)
+            {
+                size *= nitems;
+                if (size == 0)
+                {
+                    return 0;
+                }
+
+                EasyRequest easy;
+                if (TryGetEasyRequestFromContext(context, out easy))
+                {
+                    Debug.Assert(easy.RequestContentStream != null, "We should only be in the send callback if we have a request content stream");
+                    Debug.Assert(easy._associatedMultiAgent != null, "The request should be associated with a multi agent.");
+
+                    // Transfer data from the request's content stream to libcurl
+                    try
+                    {
+                        byte[] arr = easy._associatedMultiAgent._transferBuffer;
```

Comment:
> CurlSendCallback will be invoked from a diff. thread not the thread that is running Process

That's not the case.  It's always invoked from the thread calling Process, which within the call to curl_multi_perform.  This is confirmed by the FAQ (http://curl.haxx.se/docs/faq.html#Does_libcurl_use_threads):

"5.18 Does libcurl use threads?
Put simply: no, libcurl will execute in the same thread you call it in. All callbacks will be called in the same thread as the one you call libcurl in."